File: mobile/ios/ADE/Features/Profile/ProfileViewModel.swift
import SwiftUI
import Combine

/**
 * Profile ViewModel
 *
 * Manages user profile and app settings
 *
 * @author ADE iOS Team
 * @since 2026-01-24
 */

@MainActor
class ProfileViewModel: ObservableObject {
    // MARK: - Published Properties

    @Published var user: User?

    // Notifications
    @Published var pushNotificationsEnabled = true
    @Published var emailNotificationsEnabled = true
    @Published var salesAlertsEnabled = true
    @Published var lowStockAlertsEnabled = true

    // Appearance
    @Published var selectedTheme: AppTheme = .system
    @Published var selectedLanguage: AppLanguage = .turkish

    // Security
    @Published var biometricEnabled = false
    @Published var twoFactorEnabled = false

    // App Info
    let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
    let buildNumber = Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "1"

    // MARK: - Private Properties

    private let apiClient = APIClient.shared
    private var cancellables = Set<AnyCancellable>()
    private let userDefaults = UserDefaults.standard

    // Keys
    private let pushNotifKey = "push_notifications_enabled"
    private let emailNotifKey = "email_notifications_enabled"
    private let salesAlertsKey = "sales_alerts_enabled"
    private let lowStockAlertsKey = "low_stock_alerts_enabled"
    private let themeKey = "selected_theme"
    private let languageKey = "selected_language"
    private let biometricKey = "biometric_enabled"
    private let twoFactorKey = "two_factor_enabled"

    // MARK: - Initialization

    init() {
        loadSettings()
        setupObservers()
    }

    private func loadSettings() {
        // Load notification settings
        pushNotificationsEnabled = userDefaults.bool(forKey: pushNotifKey)
        emailNotificationsEnabled = userDefaults.bool(forKey: emailNotifKey)
        salesAlertsEnabled = userDefaults.bool(forKey: salesAlertsKey)
        lowStockAlertsEnabled = userDefaults.bool(forKey: lowStockAlertsKey)

        // Load appearance settings
        if let themeRaw = userDefaults.string(forKey: themeKey),
           let theme = AppTheme(rawValue: themeRaw) {
            selectedTheme = theme
        }

        if let langRaw = userDefaults.string(forKey: languageKey),
           let language = AppLanguage(rawValue: langRaw) {
            selectedLanguage = language
        }

        // Load security settings
        biometricEnabled = userDefaults.bool(forKey: biometricKey)
        twoFactorEnabled = userDefaults.bool(forKey: twoFactorKey)
    }

    private func setupObservers() {
        // Notification settings
        $pushNotificationsEnabled
            .dropFirst()
            .sink { [weak self] value in
                self?.userDefaults.set(value, forKey: self?.pushNotifKey ?? "")
                self?.updateNotificationSettings()
            }
            .store(in: &cancellables)

        $emailNotificationsEnabled
            .dropFirst()
            .sink { [weak self] value in
                self?.userDefaults.set(value, forKey: self?.emailNotifKey ?? "")
                self?.updateNotificationSettings()
            }
            .store(in: &cancellables)

        $salesAlertsEnabled
            .dropFirst()
            .sink { [weak self] value in
                self?.userDefaults.set(value, forKey: self?.salesAlertsKey ?? "")
                self?.updateNotificationSettings()
            }
            .store(in: &cancellables)

        $lowStockAlertsEnabled
            .dropFirst()
            .sink { [weak self] value in
                self?.userDefaults.set(value, forKey: self?.lowStockAlertsKey ?? "")
                self?.updateNotificationSettings()
            }
            .store(in: &cancellables)

        // Appearance settings
        $selectedTheme
            .dropFirst()
            .sink { [weak self] theme in
                self?.userDefaults.set(theme.rawValue, forKey: self?.themeKey ?? "")
                self?.applyTheme(theme)
            }
            .store(in: &cancellables)

        $selectedLanguage
            .dropFirst()
            .sink { [weak self] language in
                self?.userDefaults.set(language.rawValue, forKey: self?.languageKey ?? "")
                // Apply language change
            }
            .store(in: &cancellables)

        // Security settings
        $biometricEnabled
            .dropFirst()
            .sink { [weak self] value in
                self?.userDefaults.set(value, forKey: self?.biometricKey ?? "")
                self?.updateSecuritySettings()
            }
            .store(in: &cancellables)

        $twoFactorEnabled
            .dropFirst()
            .sink { [weak self] value in
                self?.userDefaults.set(value, forKey: self?.twoFactorKey ?? "")
                if value {
                    self?.enableTwoFactor()
                } else {
                    self?.disableTwoFactor()
                }
            }
            .store(in: &cancellables)
    }

    // MARK: - Profile Management

    func loadUserProfile() async {
        do {
            let userResponse: User = try await apiClient.request(
                endpoint: .getProfile,
                method: .get,
                requiresAuth: true
            )

            user = userResponse

        } catch {
            print("Error loading profile: \(error)")
        }
    }

    func updateProfile(name: String, email: String, phone: String, companyName: String) async {
        do {
            let request = UpdateProfileRequest(
                name: name,
                email: email,
                phone: phone,
                companyName: companyName
            )

            let updatedUser: User = try await apiClient.request(
                endpoint: .updateProfile,
                method: .put,
                body: request,
                requiresAuth: true
            )

            user = updatedUser

        } catch {
            print("Error updating profile: \(error)")
        }
    }

    // MARK: - Settings Management

    private func updateNotificationSettings() {
        Task {
            do {
                let settings = NotificationSettings(
                    pushEnabled: pushNotificationsEnabled,
                    emailEnabled: emailNotificationsEnabled,
                    salesAlerts: salesAlertsEnabled,
                    lowStockAlerts: lowStockAlertsEnabled
                )

                let _: EmptyResponse = try await apiClient.request(
                    endpoint: .updateNotificationSettings,
                    method: .put,
                    body: settings,
                    requiresAuth: true
                )

            } catch {
                print("Error updating notification settings: \(error)")
            }
        }
    }

    private func updateSecuritySettings() {
        Task {
            do {
                let settings = SecuritySettings(
                    biometricEnabled: biometricEnabled,
                    twoFactorEnabled: twoFactorEnabled
                )

                let _: EmptyResponse = try await apiClient.request(
                    endpoint: .updateSecuritySettings,
                    method: .put,
                    body: settings,
                    requiresAuth: true
                )

            } catch {
                print("Error updating security settings: \(error)")
            }
        }
    }

    private func applyTheme(_ theme: AppTheme) {
        // Apply theme to the app
        if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene {
            windowScene.windows.forEach { window in
                switch theme {
                case .light:
                    window.overrideUserInterfaceStyle = .light
                case .dark:
                    window.overrideUserInterfaceStyle = .dark
                case .system:
                    window.overrideUserInterfaceStyle = .unspecified
                }
            }
        }
    }

    // MARK: - Two-Factor Authentication

    private func enableTwoFactor() {
        Task {
            do {
                let response: TwoFactorResponse = try await apiClient.request(
                    endpoint: .enableTwoFactor,
                    method: .post,
                    requiresAuth: true
                )

                // Show QR code or secret to user
                print("Two-factor secret: \(response.secret)")

            } catch {
                print("Error enabling two-factor: \(error)")
                twoFactorEnabled = false
            }
        }
    }

    private func disableTwoFactor() {
        Task {
            do {
                let _: EmptyResponse = try await apiClient.request(
                    endpoint: .disableTwoFactor,
                    method: .post,
                    requiresAuth: true
                )

            } catch {
                print("Error disabling two-factor: \(error)")
                twoFactorEnabled = true
            }
        }
    }

    // MARK: - Account Actions

    func logout() async {
        do {
            let _: EmptyResponse = try await apiClient.request(
                endpoint: .logout,
                method: .post,
                requiresAuth: true
            )

            // Clear local data
            userDefaults.removeObject(forKey: "access_token")
            userDefaults.removeObject(forKey: "refresh_token")

            // Navigate to login screen (handled by app coordinator)
            NotificationCenter.default.post(name: .userDidLogout, object: nil)

        } catch {
            print("Error logging out: \(error)")
        }
    }

    func deleteAccount() async {
        do {
            let _: EmptyResponse = try await apiClient.request(
                endpoint: .deleteAccount,
                method: .delete,
                requiresAuth: true
            )

            // Clear all local data
            if let bundleID = Bundle.main.bundleIdentifier {
                userDefaults.removePersistentDomain(forName: bundleID)
            }

            // Navigate to login screen
            NotificationCenter.default.post(name: .userDidLogout, object: nil)

        } catch {
            print("Error deleting account: \(error)")
        }
    }

    func rateApp() {
        if let url = URL(string: "https://apps.apple.com/app/id123456789") {
            UIApplication.shared.open(url)
        }
    }
}

// MARK: - Models

struct User: Codable {
    let id: String
    let name: String
    let email: String
    let phone: String?
    let companyName: String?
    let profileImageUrl: String?
}

struct UpdateProfileRequest: Encodable {
    let name: String
    let email: String
    let phone: String
    let companyName: String
}

struct NotificationSettings: Encodable {
    let pushEnabled: Bool
    let emailEnabled: Bool
    let salesAlerts: Bool
    let lowStockAlerts: Bool
}

struct SecuritySettings: Encodable {
    let biometricEnabled: Bool
    let twoFactorEnabled: Bool
}

struct TwoFactorResponse: Decodable {
    let secret: String
    let qrCode: String
}

struct EmptyResponse: Decodable {}

// MARK: - Enums

enum AppTheme: String, CaseIterable {
    case light = "light"
    case dark = "dark"
    case system = "system"

    var displayName: String {
        switch self {
        case .light: return "Açık"
        case .dark: return "Koyu"
        case .system: return "Sistem"
        }
    }
}

enum AppLanguage: String, CaseIterable {
    case turkish = "tr"
    case english = "en"

    var displayName: String {
        switch self {
        case .turkish: return "Türkçe"
        case .english: return "English"
        }
    }
}

// MARK: - Notifications

extension Notification.Name {
    static let userDidLogout = Notification.Name("userDidLogout")
}

// MARK: - API Endpoints

extension APIEndpoint {
    static let getProfile = APIEndpoint.custom(path: "/auth/profile")
    static let updateProfile = APIEndpoint.custom(path: "/auth/profile")
    static let updateNotificationSettings = APIEndpoint.custom(path: "/settings/notifications")
    static let updateSecuritySettings = APIEndpoint.custom(path: "/settings/security")
    static let enableTwoFactor = APIEndpoint.custom(path: "/auth/2fa/enable")
    static let disableTwoFactor = APIEndpoint.custom(path: "/auth/2fa/disable")
    static let logout = APIEndpoint.custom(path: "/auth/logout")
    static let deleteAccount = APIEndpoint.custom(path: "/auth/delete-account")
}

File: mobile/android/app/src/main/java/tr/gov/ade/features/profile/ProfileViewModel.kt
package tr.gov.ade.features.profile

import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import tr.gov.ade.core.network.APIClient
import javax.inject.Inject

/**
 * Profile ViewModel
 *
 * Manages user profile and app settings
 *
 * @author ADE Android Team
 * @since 2026-01-24
 */

@HiltViewModel
class ProfileViewModel @Inject constructor(
    @ApplicationContext private val context: Context,
    private val apiClient: APIClient
) : ViewModel() {

    // MARK: - State Flows

    private val _user = MutableStateFlow<User?>(null)
    val user: StateFlow<User?> = _user.asStateFlow()

    // Notifications
    private val _pushNotificationsEnabled = MutableStateFlow(true)
    val pushNotificationsEnabled: StateFlow<Boolean> = _pushNotificationsEnabled.asStateFlow()

    private val _emailNotificationsEnabled = MutableStateFlow(true)
    val emailNotificationsEnabled: StateFlow<Boolean> = _emailNotificationsEnabled.asStateFlow()

    private val _salesAlertsEnabled = MutableStateFlow(true)
    val salesAlertsEnabled: StateFlow<Boolean> = _salesAlertsEnabled.asStateFlow()

    private val _lowStockAlertsEnabled = MutableStateFlow(true)
    val lowStockAlertsEnabled: StateFlow<Boolean> = _lowStockAlertsEnabled.asStateFlow()

    // Appearance
    private val _selectedTheme = MutableStateFlow(AppTheme.SYSTEM)
    val selectedTheme: StateFlow<AppTheme> = _selectedTheme.asStateFlow()

    private val _selectedLanguage = MutableStateFlow(AppLanguage.TURKISH)
    val selectedLanguage: StateFlow<AppLanguage> = _selectedLanguage.asStateFlow()

    // Security
    private val _biometricEnabled = MutableStateFlow(false)
    val biometricEnabled: StateFlow<Boolean> = _biometricEnabled.asStateFlow()

    private val _twoFactorEnabled = MutableStateFlow(false)
    val twoFactorEnabled: StateFlow<Boolean> = _twoFactorEnabled.asStateFlow()

    // App Info
    val appVersion: String
        get() = try {
            val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
            "${packageInfo.versionName} (${packageInfo.versionCode})"
        } catch (e: Exception) {
            "1.0.0"
        }

    // MARK: - Persistence

    private val prefs = context.getSharedPreferences("profile_settings", Context.MODE_PRIVATE)

    private val pushNotifKey = "push_notifications_enabled"
    private val emailNotifKey = "email_notifications_enabled"
    private val salesAlertsKey = "sales_alerts_enabled"
    private val lowStockAlertsKey = "low_stock_alerts_enabled"
    private val themeKey = "selected_theme"
    private val languageKey = "selected_language"
    private val biometricKey = "biometric_enabled"
    private val twoFactorKey = "two_factor_enabled"

    init {
        loadSettings()
    }

    private fun loadSettings() {
        _pushNotificationsEnabled.value = prefs.getBoolean(pushNotifKey, true)
        _emailNotificationsEnabled.value = prefs.getBoolean(emailNotifKey, true)
        _salesAlertsEnabled.value = prefs.getBoolean(salesAlertsKey, true)
        _lowStockAlertsEnabled.value = prefs.getBoolean(lowStockAlertsKey, true)

        val themeOrdinal = prefs.getInt(themeKey, AppTheme.SYSTEM.ordinal)
        _selectedTheme.value = AppTheme.values()[themeOrdinal]

        val languageOrdinal = prefs.getInt(languageKey, AppLanguage.TURKISH.ordinal)
        _selectedLanguage.value = AppLanguage.values()[languageOrdinal]

        _biometricEnabled.value = prefs.getBoolean(biometricKey, false)
        _twoFactorEnabled.value = prefs.getBoolean(twoFactorKey, false)
    }

    // MARK: - Profile Management

    fun loadUserProfile() {
        viewModelScope.launch {
            apiClient.safeApiCall {
                apiClient.service.getProfile()
            }.onSuccess { user ->
                _user.value = user
            }.onFailure { error ->
                println("Error loading profile: $error")
            }
        }
    }

    fun updateProfile(name: String, email: String, phone: String, companyName: String) {
        viewModelScope.launch {
            val request = UpdateProfileRequest(
                name = name,
                email = email,
                phone = phone,
                companyName = companyName
            )

            apiClient.safeApiCall {
                apiClient.service.updateProfile(request)
            }.onSuccess { user ->
                _user.value = user
            }.onFailure { error ->
                println("Error updating profile: $error")
            }
        }
    }

    // MARK: - Notification Settings

    fun updatePushNotifications(enabled: Boolean) {
        _pushNotificationsEnabled.value = enabled
        prefs.edit().putBoolean(pushNotifKey, enabled).apply()
        updateNotificationSettings()
    }

    fun updateEmailNotifications(enabled: Boolean) {
        _emailNotificationsEnabled.value = enabled
        prefs.edit().putBoolean(emailNotifKey, enabled).apply()
        updateNotificationSettings()
    }

    fun updateSalesAlerts(enabled: Boolean) {
        _salesAlertsEnabled.value = enabled
        prefs.edit().putBoolean(salesAlertsKey, enabled).apply()
        updateNotificationSettings()
    }

    fun updateLowStockAlerts(enabled: Boolean) {
        _lowStockAlertsEnabled.value = enabled
        prefs.edit().putBoolean(lowStockAlertsKey, enabled).apply()
        updateNotificationSettings()
    }

    private fun updateNotificationSettings() {
        viewModelScope.launch {
            val settings = NotificationSettings(
                pushEnabled = _pushNotificationsEnabled.value,
                emailEnabled = _emailNotificationsEnabled.value,
                salesAlerts = _salesAlertsEnabled.value,
                lowStockAlerts = _lowStockAlertsEnabled.value
            )

            apiClient.safeApiCall {
                apiClient.service.updateNotificationSettings(settings)
            }.onFailure { error ->
                println("Error updating notification settings: $error")
            }
        }
    }

    // MARK: - Appearance Settings

    fun updateTheme(theme: AppTheme) {
        _selectedTheme.value = theme
        prefs.edit().putInt(themeKey, theme.ordinal).apply()
        // Apply theme change
    }

    fun updateLanguage(language: AppLanguage) {
        _selectedLanguage.value = language
        prefs.edit().putInt(languageKey, language.ordinal).apply()
        // Apply language change
    }

    // MARK: - Security Settings

    fun updateBiometric(enabled: Boolean) {
        _biometricEnabled.value = enabled
        prefs.edit().putBoolean(biometricKey, enabled).apply()
        updateSecuritySettings()
    }

    fun updateTwoFactor(enabled: Boolean) {
        _twoFactorEnabled.value = enabled
        prefs.edit().putBoolean(twoFactorKey, enabled).apply()

        if (enabled) {
            enableTwoFactor()
        } else {
            disableTwoFactor()
        }
    }

    private fun updateSecuritySettings() {
        viewModelScope.launch {
            val settings = SecuritySettings(
                biometricEnabled = _biometricEnabled.value,
                twoFactorEnabled = _twoFactorEnabled.value
            )

            apiClient.safeApiCall {
                apiClient.service.updateSecuritySettings(settings)
            }.onFailure { error ->
                println("Error updating security settings: $error")
            }
        }
    }

    private fun enableTwoFactor() {
        viewModelScope.launch {
            apiClient.safeApiCall {
                apiClient.service.enableTwoFactor()
            }.onSuccess { response ->
                println("Two-factor enabled: ${response.secret}")
            }.onFailure { error ->
                println("Error enabling two-factor: $error")
                _twoFactorEnabled.value = false
                prefs.edit().putBoolean(twoFactorKey, false).apply()
            }
        }
    }

    private fun disableTwoFactor() {
        viewModelScope.launch {
            apiClient.safeApiCall {
                apiClient.service.disableTwoFactor()
            }.onFailure { error ->
                println("Error disabling two-factor: $error")
                _twoFactorEnabled.value = true
                prefs.edit().putBoolean(twoFactorKey, true).apply()
            }
        }
    }

    // MARK: - Account Actions

    fun logout() {
        viewModelScope.launch {
            apiClient.safeApiCall {
                apiClient.service.logout()
            }.onSuccess {
                // Clear local data
                prefs.edit().clear().apply()
                context.getSharedPreferences("auth", Context.MODE_PRIVATE)
                    .edit().clear().apply()

                // Navigate to login (handled by navigation)
                // TODO: Trigger navigation event
            }.onFailure { error ->
                println("Error logging out: $error")
            }
        }
    }

    fun deleteAccount() {
        viewModelScope.launch {
            apiClient.safeApiCall {
                apiClient.service.deleteAccount()
            }.onSuccess {
                // Clear all local data
                prefs.edit().clear().apply()
                context.getSharedPreferences("auth", Context.MODE_PRIVATE)
                    .edit().clear().apply()

                // Navigate to login (handled by navigation)
                // TODO: Trigger navigation event
            }.onFailure { error ->
                println("Error deleting account: $error")
            }
        }
    }

    fun rateApp() {
        try {
            val intent = Intent(Intent.ACTION_VIEW).apply {
                data = Uri.parse("market://details?id=${context.packageName}")
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(intent)
        } catch (e: Exception) {
            // Fallback to browser
            val intent = Intent(Intent.ACTION_VIEW).apply {
                data = Uri.parse("https://play.google.com/store/apps/details?id=${context.packageName}")
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(intent)
        }
    }
}

// MARK: - Request Models

data class UpdateProfileRequest(
    val name: String,
    val email: String,
    val phone: String,
    val companyName: String
)

data class NotificationSettings(
    val pushEnabled: Boolean,
    val emailEnabled: Boolean,
    val salesAlerts: Boolean,
    val lowStockAlerts: Boolean
)

data class SecuritySettings(
    val biometricEnabled: Boolean,
    val twoFactorEnabled: Boolean
)

// MARK: - Response Models

data class TwoFactorResponse(
    val secret: String,
    val qrCode: String
)

// MARK: - API Service Extensions

suspend fun tr.gov.ade.core.network.APIService.getProfile(): User {
    // Placeholder implementation
    return User(
        id = "1",
        name = "Kullanıcı",
        email = "user@example.com",
        phone = null,
        companyName = null,
        profileImageUrl = null
    )
}

suspend fun tr.gov.ade.core.network.APIService.updateProfile(request: UpdateProfileRequest): User {
    // Placeholder implementation
    return User(
        id = "1",
        name = request.name,
        email = request.email,
        phone = request.phone,
        companyName = request.companyName,
        profileImageUrl = null
    )
}

suspend fun tr.gov.ade.core.network.APIService.updateNotificationSettings(settings: NotificationSettings) {
    // Placeholder implementation
}

suspend fun tr.gov.ade.core.network.APIService.updateSecuritySettings(settings: SecuritySettings) {
    // Placeholder implementation
}

suspend fun tr.gov.ade.core.network.APIService.enableTwoFactor(): TwoFactorResponse {
    // Placeholder implementation
    return TwoFactorResponse(secret = "SECRET", qrCode = "QR_CODE")
}

suspend fun tr.gov.ade.core.network.APIService.disableTwoFactor() {
    // Placeholder implementation
}

suspend fun tr.gov.ade.core.network.APIService.logout() {
    // Placeholder implementation
}

suspend fun tr.gov.ade.core.network.APIService.deleteAccount() {
    // Placeholder implementation
}
